---
title: "How does harf work?"
author: "Cesaire Fouodo"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{How does harf work?}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 400px;
}
```

## Introduction

The R package **harf** is an extension of the adversarial random forests (ARFs) 
to high-dimensional data. While in high-dimensional settings, the feature distributions 
are not necessarily independent within terminal nodes, **harf** identifies isolated regions of the feature space in order to better preserve the independence within terminal nodes. This vignette serves as a user guide to use the package effectively.
Three key functionalities are provided: `h_arf` to train a high-dimensional adversarial 
random forest (HARF) and `h_forge` for the synthetic  data generating process. This function 
allows for conditional data generation as well.

## Libraries and data

The data used in this vignette are included in the *harf* package and originate from The Cancer Genome Atlas (TCGA) and the Genotype-Tissue Expression (GTEx) projects, two large-scale public resources providing extensive RNA-seq data. The data were extracted from previously processed and curated datasets generated using the pipeline described in Aguet et al. (2017).

The built-in dataset `single_cell`, distributed with *harf*, contains $233$ log-transformed gene expression measurements from $654$ cells collected across $15$ human tissues, including bladder, breast, cervix, colon, esophagus (gastroesophageal junction, mucosa, muscularis), kidney, liver, lung, prostate, salivary gland, stomach, thyroid, and uterus. The variable `cell_type` indicates the tissue of origin for each cell.

In genetic epidemiology studies, molecular (omics) data are typically accompanied by clinical or phenotypic variables that provide essential contextual information. The HARF framework assumes that input data consist of two components: (i) high-dimensional numeric omics features and (ii) associated clinical or phenotypic variables, which may be either categorical or numeric. In this vignette, gene expression measurements are used as the omics data, while cell type serves as the clinical or phenotypic variable.

We start by installing the required packages if you have not already done so.

```{r install_libraries, warning = FALSE, message = FALSE, eval = FALSE}
install.packages("data.table")
install.packages("rsvd")
install.packages("Rtsne")
install.packages("cowplot")
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("SingleCellExperiment")
BiocManager::install("scater")
install.packages("ggplot2")
install.packages("corrplot")
```

We load the libraries.

```{r load_libraries, warning = FALSE, message = FALSE}
library(harf)
library(data.table)
library(rsvd)
library(Rtsne)
library(cowplot)
library(SingleCellExperiment)
library(ggplot2)
library(corrplot)
library(scater)
```

We load the single_cell dataset and disable parallelization when the code is 
executed in a `CRAN` environment, in accordance with `CRAN` policies.

```{r data_example, include=TRUE, eval=TRUE, message=FALSE}
data("single_cell")
# Do not parellelize on CRAN
parallel <- ifelse(Sys.getenv("NOT_CRAN") == "true", TRUE, FALSE)
```

## Training a harf model #

We train a HARF model using the `h_arf` function. The gene expression measurements 
are provided as the `omx_data` argument, while the cell type information is passed as 
the `cli_lab_data` argument. we use the default number of trees and other 
default random forest hyperparameters as specified in the `R` package `ranger`.

```{r harf_training, include = TRUE, eval = TRUE, message=FALSE}
set.seed(123)
harf_model <- h_arf(
 omx_data = single_cell[ , - which(colnames(single_cell)  == "cell_type")],
 cli_lab_data = data.frame(cell_type = single_cell$cell_type),
 parallel = parallel,
 verbose = FALSE
)
```

## Generating synthetic data

The `h_forge` function generates synthetic data from a trained HARF model. Here, 
we generate the same number of synthetic samples as in the original dataset, and 
without any evidence (i.e., unconditional generation).

```{r harf_synthetic_data, include = TRUE, eval = TRUE, message=FALSE}
set.seed(123)
synth_single_cell <- h_forge(
  harf_obj = harf_model,
  n_synth = nrow(single_cell),
  evidence = NULL,
  parallel = parallel,
  verbose = FALSE
  )
```

## Conditional expectations

For conditional data generation, we generate synthetic samples for each cell type 
separately. The generated samples are then combined to form the final synthetic
dataset. Note that our current actual `harf` object allows for conditioning on clinical 
feature only. 

```{r harf_conditional_expectation, include = TRUE, eval = TRUE, message=FALSE}
set.seed(123)
single_cell_list <- lapply(unique(single_cell$cell_type), function (ct) {
  ct_synth <- h_forge(
        harf_obj = harf_model,
        n_synth = sum(single_cell$cell_type == ct),
        evidence = data.frame(cell_type = ct),
        verbose = FALSE,
        parallel = parallel
      )
  return(ct_synth)
})
cond_synth_single_cell <- do.call(rbind, single_cell_list)
```

## Comparaison of correlation matrices

We visually compare the correlation matrices of the original data, the synthetic data, 
and the conditionally resampled data. To enhance interpretability, we rearrange the features 
according to their cluster assignments obtained from the HARF model. The correlation 
plots exhibit similar patterns across the three datasets, indicating that the HARF model 
effectively preserve the origin feature correlation structure.

```{r harf_correlation_matrices, include = TRUE, eval = TRUE, message = FALSE, fig.width = 7, fig.height = 6.5}
# Re-arrange data by grouping gene by clusters
cluster_feature <- copy(harf_model$cluster)
setorder(cluster_feature, cluster)
orig_clustered <- single_cell[ , c("cell_type", cluster_feature$feature)]
synth_clustered <- as.data.frame(synth_single_cell)[ , c("cell_type", cluster_feature$feature)]
cond_synth_clustered <- as.data.frame(cond_synth_single_cell)[ , c("cell_type", cluster_feature$feature)]
plot_corr <- function(dt, title) {
  corr_matrix <- cor(dt[ , 2:51], method = "spearman")
  corrplot(corr_matrix,
           method = "circle",
           tl.col = "black",
           tl.pos = "n",
           # tl.srt = 45,
           title = title,
           mar = c(0, 0, 1, 0))
}
par(mfrow = c(2,2))
plot_corr(orig_clustered, "Original")
plot_corr(synth_clustered, "Synthetic")
plot_corr(cond_synth_clustered, "Conditional resampling")
par(mfrow = c(1, 1))
```

## Show original and synthetic data in 2D using t-SNE

We use t-SNE to visualize the original data, the synthetic data, and the 
conditionally resampled data in a two-dimensional space. We observe that the 
three datasets exhibit similar cluster structure of tissues.

```{r harf_tsne, include = TRUE, eval = TRUE, message = FALSE, fig.width = 7.1, fig.height = 3.5}
set.seed(123)
tsne_it <- function (sc_data, perp = 30, title = "") {
  # Create SingleCellExperiment object
  sce <- SingleCellExperiment::SingleCellExperiment(
    assays = list(counts = t(as.matrix(sc_data[ , - which(colnames(sc_data)  == "cell_type")])))
  )
  SingleCellExperiment::logcounts(sce) <- SingleCellExperiment::counts(sce) # Log-normalization
  sce$cell_type <- sc_data$cell_type
  pc_sce <- rpca(t(SingleCellExperiment::counts(sce)))
  # tSNE with rotated pcs
  ts_sce <- Rtsne::Rtsne(
    pc_sce$x %*% pc_sce$rotation,
    perplexity = perp,
    verb = FALSE,
    pca = FALSE,
    check_duplicates = FALSE
  )
  SingleCellExperiment::reducedDim(sce, "tsne") = ts_sce$Y
  sce_plot <- scater::plotReducedDim(sce, "tsne", colour_by = "cell_type") + 
  ggplot2::ggtitle(title) +
  ggplot2::theme(legend.position = "bottom")
  return(sce_plot)
}  
orig_plot <- tsne_it(single_cell,
                     perp = 30,
                     title = "Original")
synth_plot <- tsne_it(as.data.frame(synth_single_cell),
                      perp = 30,
                      title = "Synthetic")
cond_synth_plot <- tsne_it(as.data.frame(cond_synth_single_cell),
                           perp = 30,
                           title = "Conditional resampling")
legend <- cowplot::get_legend(
  orig_plot + theme(legend.position = "bottom")
)
orig_plot <- orig_plot + theme(legend.position = "none")
synth_plot <- synth_plot + theme(legend.position = "none")
cond_synth_plot <- cond_synth_plot + theme(legend.position = "none")
all_plots <- cowplot::plot_grid(orig_plot, synth_plot, cond_synth_plot, ncol = 3)
plot_grid(
  all_plots,
  legend,
  ncol = 1,
  rel_heights = c(1, 0.2)
)
```

## Conclusion

We introduced the R package **harf**, which enables the generation of high-dimensional synthetic data. The package extends the adversarial random forest framework to effectively handle high-dimensional omics data and supports both unconditional and conditional data generation. Using visual comparisons based on correlation matrices and `t-SNE` projections, we showed that the synthetic data produced by **harf** closely resemble the original data in terms of feature correlations and underlying cluster structure.

While **harf** enables the simultaneous generation of both omics and clinical data, it requires that features be clustered prior to training the HARF model. This step may be challenging in situations where features do not naturally form clusters or when the optimal number of clusters or feature groups is not easily determined. Future work could explore adaptive clustering strategies or methods that relax this requirement to further improve usability in diverse high-dimensional settings.

## References

- The Cancer Genome Atlas Pan-Cancer analysis project. Nature Genetics, 45, 1113–1120 (2013). Link [here](https://www.nature.com/articles/ng.2764).

- The Genotype-Tissue Expression (GTEx) project. Nature Genetics, 47, 1231–1237. (2015). Link [here](https://www.nature.com/articles/ng.2653).

- Genetic effects on gene expression across human tissues. Nature, 550, 204–213. (2017). Link [here](https://www.nature.com/articles/nature24277).

- Q. Wang, J Armenia, C. Zhang, A.V. Penson, E. Reznik, L. Zhang, T. Minet, A. Ochoa, B.E. Gross, C. A. Iacobuzio-Donahue, D. Betel, B.S. Taylor, J. Gao, N. Schultz. Unifying cancer and normal RNA sequencing data from different sources. Scientific Data 5:180061, 2018. Link [here](https://www.nature.com/articles/sdata201861).

- Fouodo, C. J. K., et al. (2026). High-dimensional adversarial random forests. Submission. Link [don't click](https://arxiv.org/abs/2405.12345).

- Watson, D. S., Blesch, K., Kapar, J. & Wright, M. N. (2023). Adversarial random forests for density estimation and generative modeling. In Proceedings of the 26th International Conference on Artificial Intelligence and Statistics. Link [here](https://proceedings.mlr.press/v206/watson23a.html).
